### PR TITLE
Fix passing object ids in local mode

### LINKED ISSUE #6115: Passing ObjectID as a function argument in local_mode is broken
### System information
- **OS Platform and Distribution**: Ubuntu 18.04
- **Ray installed from (source or binary)**: binary
- **Ray version**: 0.8.0.dev6
- **Python version**: 3.7
- **Exact command to reproduce**: see below

### Describe the problem
The argument passing behavior with local_mode=True vs False seems to be different.
When I run the code snippet below:

```import ray
ray.init(local_mode=True)  # Replace with False to get a working example

@ray.remote
def remote_function(x):
    obj = x['a']
    return ray.get(obj)


a = ray.put(42)
d = {'a': a}
result = remote_function.remote(d)
print(ray.get(result))
```

With local_mode=False I get output `42`, as expected.
With local_mode=True I get the following error:

```
Traceback (most recent call last):
  File "/home/alex/all/projects/doom-neurobot/playground/ray_local_mode_bug.py", line 13, in <module>
    print(ray.get(result))
  File "/home/alex/miniconda3/envs/doom-rl/lib/python3.7/site-packages/ray/worker.py", line 2194, in get
    raise value.as_instanceof_cause()
ray.exceptions.RayTaskError(KeyError): /home/alex/miniconda3/envs/doom-rl/bin/python /home/alex/all/projects/doom-neurobot/playground/ray_local_mode_bug.py (pid=2449, ip=10.136.109.38)
  File "/home/alex/miniconda3/envs/doom-rl/lib/python3.7/site-packages/ray/local_mode_manager.py", line 55, in execute
    results = function(*copy.deepcopy(args))
  File "/home/alex/all/projects/doom-neurobot/playground/ray_local_mode_bug.py", line 7, in remote_function
    return ray.get(obj)
  File "/home/alex/miniconda3/envs/doom-rl/lib/python3.7/site-packages/ray/local_mode_manager.py", line 105, in get_objects
    raise KeyError("Value for {} not found".format(object_id))
KeyError: 'Value for LocalModeObjectID(89f92e430883458c8107c10ed53eb35b26099831) not found'
```
It looks like the LocalObjectID instance inside `d` loses it's field `value` when it gets deep copied during the "remote" function call (currently it's `local_mode_manager.py:55`). It's hard to tell why exactly that happens, looks like a bug.

### PR DESCRIPTION
**Pull request by edoakes**
## Related issue number

Closes #6115

## Checks

- [x] I've run `scripts/format.sh` to lint the changes in this PR.
- [x] I've included any doc changes needed for https://ray.readthedocs.io/en/latest/.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failure rates at https://ray-travis-tracker.herokuapp.com/.

### CODE REVIEW

File: python/ray/local_mode_manager.py [hunk 1]
```diff
@@ -20,7 +21,18 @@ class LocalModeObjectID(ObjectID):
...
     """
-    pass
+
+    def __copy__(self):
+        new = LocalModeObjectID(self.binary())
+        if hasattr(self, "value"):
+            new.value = self.value
+        return new
+
+    def __deepcopy__(self, memo={}):
```
Comment (richardliaw) [inline python/ray/local_mode_manager.py]: can you take out the dict? https://docs.python-guide.org/writing/gotchas/#mutable-default-arguments

File: python/ray/local_mode_manager.py [hunk 2]
```diff
@@ -49,23 +61,37 @@ def execute(self, function, function_name, args, kwargs, num_return_vals):
...
-        object_ids = [
+        return_ids = [
             LocalModeObjectID.from_random() for _ in range(num_return_vals)
         ]
+        new_args = []
+        for i, arg in enumerate(args):
+            if isinstance(arg, ObjectID):
+                new_args.append(ray.get(arg))
+            else:
+                new_args.append(copy.deepcopy(arg))
```
Comment (richardliaw) [inline python/ray/local_mode_manager.py]: does it make sense to push this through plasma serialization? would that make this more realistic?

**Review by richardliaw — APPROVED** [review]
approve conditionally on memo fix

File: python/ray/local_mode_manager.py (hunk 2 above)
Comment (edoakes) [reply python/ray/local_mode_manager.py]: yes but this is not totally trivial so let's leave it to another patch

**Pull request merged**